### ISSUE TITLE
Nmap client certificate authentication not working

### ISSUE THREAD

**Issue by vigneshkaladharan** [nmap/nmap#2059]
Hi,
    I have installed nmap 7.8 and run the following command 

**nmap host  -T3 --script +ssl-cert -p2433**

The server requires client cert for authentication 

NSE: Starting ssl-cert M:251cec8 against 192.168.150.40:5200.
NSOCK INFO [0.5310s] nsock_iod_new2(): nsock_iod_new (IOD #1)
NSOCK INFO [0.5910s] nsock_connect_ssl(): SSL connection requested to 192.168.150.40:5200/tcp (IOD #1) EID 9
NSOCK INFO [0.8800s] handle_connect_result(): EID 9 reconnecting with SSL_OP_NO_SSLv2
NSOCK INFO [1.1260s] handle_connect_result(): EID 9 error:14094412:SSL routines:ssl3_read_bytes:sslv3 alert bad certificate
NSOCK INFO [1.1260s] nsock_trace_handler_callback(): Callback: SSL-CONNECT ERROR [Input/output error (5)] for EID 9 [192.168.150.40:5200]
NSE: [ssl-cert M:251cec8 192.168.150.40:5200] getCertificate error: Failed to connect to server
NSE: Finished ssl-cert M:251cec8 against 192.168.150.40:5200.
NSOCK INFO [1.1260s] nsock_iod_delete(): nsock_iod_delete (IOD #1)
NSOCK INFO [1.1260s] nsock_iod_delete(): nsock_iod_delete: SSL shutdown failed ((null)) on NSI 1
Completed NSE at 14:29, 0.54s elapsed
Nmap scan report for 192.168.150.40
Host is up, received conn-refused (0.22s latency).
Scanned at 2020-06-08 14:29:46 IST for 0s

Do i need to provide any attributes for this

**Comment by cnotin** [comment]
Can you please detail what did you expect instead?
That Nmap would collect the server certificate anyway (in that case it's related to #1995 no need to create another) or that it would authenticate (but I'm not aware that Nmap supports it) and do something?

**Comment by vigneshkaladharan** [comment]
I want to collect ssl certificate from a host.
when i run nmap host -T3 --script +ssl-cert -p2433 to collect server certificate from the host,
i am getting an error since the host performs client certificate authentication.

wanted to know is there provisions in nmap to provide client certificate when we run nmap commands.

**Comment by cnotin** [comment]
I'm not aware of Nmap providing client certificate authentication. But the thing is that we get the server certificate before authenticating so it's actually not necessary to authenticate. But the current code doesn't support it.

**Comment by vigneshkaladharan** [comment]
thanks for your reply

i am facing like #1488 ,u have mentioned you were able to retrieve the server certificate even with client cert authentication failure.

do i need to apply any patch on 7.8 version

**Comment by cnotin** [comment]
I mean technically Nmap could retrieve the server certificate. Check with a packet capture (wireshark) in parallel and you'll see it in the "server hello". But currently no one has suggested code to implement it in nmap itself

**Comment by cnotin** [comment]
(by the way, latest version is 7.80 and not 7.8 😉)

**Comment by vigneshkaladharan** [comment]
Yes i do use 7.80 :)

**Comment by vigneshkaladharan** [comment]
> I mean technically Nmap could retrieve the server certificate. Check with a packet capture (wireshark) in parallel and you'll see it in the "server hello". But currently no one has suggested code to implement it in nmap itself

which mean you could not see the certificate in ouputstream

**Comment by cnotin** [comment]
No currently nmap won't show you the server cert when client certificate auth is required. But you can still see it in the packet capture if you really need it.

**Comment by vigneshkaladharan** [comment]
okie @cnotin  thank you

**Comment by git-blame** [comment]
Using nmap 7.4 and 7.7 I'm able to connect to sites (cryptomix, badssl) requiring client auth as documented [here](https://stackoverflow.com/questions/38095559/https-test-server-that-checks-client-certificates) and nmap returns the server cert.

Looking over your logs, I don't think it's client auth but maybe a malformed server cert that's your problem?
```
NSOCK INFO [0.8800s] handle_connect_result(): EID 9 reconnecting with SSL_OP_NO_SSLv2
NSOCK INFO [1.1260s] handle_connect_result(): EID 9 error:14094412:SSL routines:ssl3_read_bytes:sslv3 alert bad certificate
NSOCK INFO [1.1260s] nsock_trace_handler_callback(): Callback: SSL-CONNECT ERROR 
```

The script _ssl-cert_ relies on the script _sslcert_ to return the cert. The latter tries to use openssl if available, else its own pseudo handshake which is just enough to retrieve the cert. So you could try to set _have_openssl = false_ in that script and see what happens.

```lua
function getCertificate(host, port)
  -- HACK: set to false so we don't use openssl
  have_openssl = false 
  local mutex = nmap.mutex("sslcert-cache-mutex")
  mutex "lock"
```

**Comment by cnotin** [comment]
Thanks I'll check again with this!

**Comment by cnotin** [comment]
Oops sorry everyone I confused this issue with #1995 
Nmap collects fine the server certificate even when client certificate is required:
```console
# ./nmap --script +ssl-cert -p443 client.badssl.com
Starting Nmap 7.80SVN ( https://nmap.org ) at 2020-06-21 19:15 CEST
Nmap scan report for client.badssl.com (104.154.89.105)
Host is up (0.026s latency).
rDNS record for 104.154.89.105: 105.89.154.104.bc.googleusercontent.com

PORT    STATE SERVICE
443/tcp open  https
| ssl-cert: Subject: commonName=*.badssl.com/organizationName=Lucas Garron Torres/stateOrProvinceName=California/countryName=US
| Subject Alternative Name: DNS:*.badssl.com, DNS:badssl.com
| Issuer: commonName=DigiCert SHA2 Secure Server CA/organizationName=DigiCert Inc/countryName=US
| Public Key type: rsa
| Public Key bits: 2048
| Signature Algorithm: sha256WithRSAEncryption
| Not valid before: 2020-03-23T00:00:00
| Not valid after:  2022-05-17T12:00:00
| MD5:   d645 2e2f f0ac 029e 1411 1204 8e6c 660b
|_SHA-1: 1845 b216 efd0 839a 1851 a957 325d a336 2170 49cb

Nmap done: 1 IP address (1 host up) scanned in 1.28 seconds
```

**Comment by vigneshkaladharan** [comment]
> Oops sorry everyone I confused this issue with #1995
> Nmap collects fine the server certificate even when client certificate is required:
> 
> ```
> # ./nmap --script +ssl-cert -p443 client.badssl.com
> Starting Nmap 7.80SVN ( https://nmap.org ) at 2020-06-21 19:15 CEST
> Nmap scan report for client.badssl.com (104.154.89.105)
> Host is up (0.026s latency).
> rDNS record for 104.154.89.105: 105.89.154.104.bc.googleusercontent.com
> 
> PORT    STATE SERVICE
> 443/tcp open  https
> | ssl-cert: Subject: commonName=*.badssl.com/organizationName=Lucas Garron Torres/stateOrProvinceName=California/countryName=US
> | Subject Alternative Name: DNS:*.badssl.com, DNS:badssl.com
> | Issuer: commonName=DigiCert SHA2 Secure Server CA/organizationName=DigiCert Inc/countryName=US
> | Public Key type: rsa
> | Public Key bits: 2048
> | Signature Algorithm: sha256WithRSAEncryption
> | Not valid before: 2020-03-23T00:00:00
> | Not valid after:  2022-05-17T12:00:00
> | MD5:   d645 2e2f f0ac 029e 1411 1204 8e6c 660b
> |_SHA-1: 1845 b216 efd0 839a 1851 a957 325d a336 2170 49cb
> 
> Nmap done: 1 IP address (1 host up) scanned in 1.28 seconds
> ```

yes @cnotin @git-blame its working fine.seems problem with our server configuration.
Thanks for your valuable reply.

**Comment by fyodor** [comment]
Thanks all.  Sounds like this was figured out.
